File: Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left, right = 1, 1
        left_p = [1] * len(nums)
        right_p = [1] * len(nums)
        len_nums = len(nums)

        for i in range(1, len_nums):
            left = left * nums[i - 1]
            left_p[i] = left

            right = right * nums[len_nums - i]
            right_p[len_nums - i - 1] = right

        for i in range(len_nums):
            nums[i] = left_p[i] * right_p[i]

        return nums

File: java_version/src/ProductofArrayExceptSelf.java
public class ProductofArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] left = new int[n];
        int[] right = new int[n];
        left[0] = 1;
        right[n-1] = 1;
        for (int i=1; i<n; i++){
            left[i] = left[i-1]*nums[i-1];
            right[n-i-1] = right[n-i]*nums[n-i];
        }
        for(int i=0; i< n; i++)
            nums[i] = left[i]*right[i];
        return nums;
    }
}
